src/iop/lut3d.c

```diff
@@ -0,0 +1,1071 @@
+    }
+    else // file chosen outside of root folder
+    {
+      fprintf(stderr, "[lut3d] Select file outside Lut root folder is not allowed\n");
+      dt_control_log(_("Select file outside Lut root folder is not allowed"));
+      g_free(lutfolder);
+      gtk_widget_destroy(filechooser);
+      return;
+    }
+    gtk_entry_set_text(GTK_ENTRY(g->filepath), filepath);
+    g_free(filepath);
+  }
+  g_free(lutfolder);
+  gtk_widget_destroy(filechooser);
+}
+
+void gui_reset(dt_iop_module_t *self)
+{
+  memcpy(self->params, self->default_params, sizeof(dt_iop_lut3d_params_t));
+}
+
+void gui_update(dt_iop_module_t *self)
+{
+  dt_iop_lut3d_gui_data_t *g = (dt_iop_lut3d_gui_data_t *)self->gui_data;
+  dt_iop_lut3d_params_t *p = (dt_iop_lut3d_params_t *)self->params;
+  gchar *lutfolder = dt_conf_get_string("plugins/darkroom/lut3d/def_path");
+  gtk_entry_set_text(GTK_ENTRY(g->filepath), p->filepath);
+  if (lutfolder[0] == 0)
+  {
+    gtk_widget_set_sensitive(g->hbox, FALSE);
+  }
+  else
+  {
+    gtk_widget_set_sensitive(g->hbox, TRUE);
+  }
+  g_free(lutfolder);
+  dt_bauhaus_combobox_set(g->colorspace, p->colorspace);
+  dt_bauhaus_combobox_set(g->interpolation, p->interpolation);
+}
+
+void gui_init(dt_iop_module_t *self)
+{
+  self->gui_data = malloc(sizeof(dt_iop_lut3d_gui_data_t));
+  dt_iop_lut3d_gui_data_t *g = (dt_iop_lut3d_gui_data_t *)self->gui_data;
+
+  self->widget = gtk_box_new(GTK_ORIENTATION_VERTICAL, DT_BAUHAUS_SPACE);
+  dt_gui_add_help_link(self->widget, dt_get_help_url(self->op));
+
+  g->hbox = gtk_box_new(GTK_ORIENTATION_HORIZONTAL, DT_PIXEL_APPLY_DPI(8));
+  g->filepath = gtk_entry_new();
+  gtk_box_pack_start(GTK_BOX(g->hbox), g->filepath, TRUE, TRUE, 0);
+  dt_gui_key_accel_block_on_focus_connect(GTK_WIDGET(g->filepath));
+  gtk_widget_set_tooltip_text(g->filepath, _("the file path (relative to lut folder) is saved with image (and not the lut data themselves)\n"
+                                              "CAUTION: lut folder must be set in preferences/core options/miscellaneous before choosing the lut file"));
+  g_signal_connect(G_OBJECT(g->filepath), "changed", G_CALLBACK(filepath_callback), self);
+
+  GtkWidget *button = dtgtk_button_new(dtgtk_cairo_paint_directory, CPF_DO_NOT_USE_BORDER, NULL);
+  gtk_widget_set_size_request(button, DT_PIXEL_APPLY_DPI(18), DT_PIXEL_APPLY_DPI(18));
+  gtk_widget_set_tooltip_text(button, _("select a png (haldclut) or a cube file"));
+  gtk_box_pack_start(GTK_BOX(g->hbox), button, FALSE, FALSE, 0);
+  g_signal_connect(G_OBJECT(button), "clicked", G_CALLBACK(button_clicked), self);
+
+  gtk_box_pack_start(GTK_BOX(self->widget), GTK_WIDGET(g->hbox), TRUE, TRUE, 0);
+
+  g->colorspace = dt_bauhaus_combobox_new(self);
+  dt_bauhaus_widget_set_label(g->colorspace, NULL, _("application color space"));
+  dt_bauhaus_combobox_add(g->colorspace, _("sRGB"));
+  dt_bauhaus_combobox_add(g->colorspace, _("REC.709"));
+  dt_bauhaus_combobox_add(g->colorspace, _("lin sRGB/REC.709"));

```

Comment:
"lin" here stand for linear?

Can't we rename this to "sRGB/REC.709 (linear)" ?

Likewise for the next line. What do you think?

Reply:
Yes, it would better with full words: 
```
linear sRGB/REC.709
linear prophoto RGB
```
Another change I'd like to make. At the last minute I've replaced `copyright (c) 2016 Chih-Mao Chen` with my name. I should have kept this line and added another one for myself.

Do I issue a new PR ? Do we wait for more comments ?

PS: When I've rebased last time I had copied the full lut3d.c file into another editor and pasted it again to get back my last changes. That's probably why you have found unexpected crlf into the file.

Reply:
Please be careful with names. sRGB is REC709 primaries + a piece-wise "gamma". There is no linear sRGB. Linear REC709 is REC709 primaries + nothing else. These are standards with precise meanings, it's bad to mess around with their names.

Reply:
If I understand properly we end up with:
_sRGB_ // Linear REC709 + gamma
_REC.709_ // Linear REC709 + gamma (OETF)
_linear REC.709_
_linear prophoto RGB_

It could be useful to provide the source of technical vocabulary we refer to (I've to admit I've read everything and its opposite ...)

Reply:
Ok, to be super strict on the terms:
1. only displays (screens) have a gamma, so every transfer function built into a color space is an OETF (or EOTF, depending if it's an input or output space). What we call gamma for RGB spaces is an abuse and should be called OETF.
2. a color space is defined by RGB primaries (expressed as subvectors of the XYZ space) AND an OETF/EOTF.

Because video cameras don't record raw signals but use an encoding for the sake of storage space, they output a non-linear signal with a built-in OETF. So, to apply a LUT intended for video camera to photographs coming out of a DSLR, we need to convert from camera RGB to linear REC709, then apply a the same OETF (only for the sake of compatibility of the LUT) as the camera one.

Such OETFs could be the Rec.UIT-R BT.709-6 OETF, but could also be any log curve (S-log, N-log, etc.) on top of the Rec709 primaries (as far as I understand).

But sRGB is an output space, for display (not used in recording), so it's "gamma" is an EOTF. I believe it's useful for us only for compatibility with HaldCLUTs.

So, the options we have are :

* sRGB
* Rec.709 + HDTV OETF
* Linear Rec.709
* Linear ProphotoRGB (why this one ?)

Reply:
I follow you on OETF and EOTF (sRGB or Rec709) ends. If Rec709 primaries are used in the middle for the TV world (today), they are not the only ones, Rec2020 can be used as well. On the other hand, the log raw has been introduced to improve the color accuracy in the shadows (out of camera) but I haven't found anything saying it refers specifically to Rec709.

Practically I haven't found any effective color grading lut working in linear Rec709 (in linear prophoto RGB neither). I've let these color space for testing purpose but until we find some luts working with linear spaces we can remove them.

Another point could come in the future. There is no reason to have always input = output color space for the lut transform. We could have a lut working on linear Rec709 and producing an image in Rec709 EOTF for example. But again I haven't not found any reference about this. I've made some tests during the development but without convincing results. 

To keep it simple for the time being, we could just keep:
- sRGB
- Rec.709
both display referred.
In the future we could add the log ones (if we get camera log raw) and the linear ones (if useful). We may also offer to select a different ouput space if the need shows up.